### PR TITLE
Refactor CommonCompiler.RunCore

### PR DESCRIPTION
**Pull request by agocke**
This is the monolith method for batch compilation in C# and VB.
It is large and confusing. I've tried to make it better by moving to a single
error reporting system for every piece of the pipeline (Diagnostics in DiagnosticBags).

It may be helpful to look at each commit separately.

### CODE REVIEW

**Comment by agocke** [comment]
ping @dotnet/roslyn-compiler for review

File: src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs [hunk 1]
```diff
@@ -429,6 +429,10 @@ public bool ReportErrors(IEnumerable<Diagnostic> diagnostics, TextWriter console
             return hasErrors;
         }
 
+        private bool ReportErrors(DiagnosticBag diagnostics, TextWriter consoleOutput, ErrorLogger errorLoggerOpt)
+            => ReportErrors(diagnostics.ToReadOnly(), consoleOutput, errorLoggerOpt);
```
Comment (jcouv) [inline src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs]: Nit: two empty lines

File: src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs [hunk 2]
```diff
@@ -385,7 +385,7 @@ internal static DiagnosticInfo ToFileReadDiagnostics(CommonMessageProvider messa
             return diagnosticInfo;
         }
 
-        public bool ReportErrors(IEnumerable<Diagnostic> diagnostics, TextWriter consoleOutput, ErrorLogger errorLoggerOpt)
+        internal bool ReportErrors(IEnumerable<Diagnostic> diagnostics, TextWriter consoleOutput, ErrorLogger errorLoggerOpt)
```
Comment (jcouv) [inline src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs]: Could IEnumerable be switched to ImmutableArray, or are there callers that still need IEnumerable?

File: src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs [hunk 3]
```diff
@@ -559,304 +571,337 @@ private int RunCore(TextWriter consoleOutput, ErrorLogger errorLogger, Cancellat
...
+            DiagnosticBag diagnostics,
+            CancellationToken cancellationToken,
+            out CancellationTokenSource analyzerCts,
+            out bool reportAnalyzer,
+            out AnalyzerDriver analyzerDriver)
+        {
+            analyzerCts = null;
+            reportAnalyzer = false;
+            analyzerDriver = null;
+            AnalyzerManager analyzerManager = null;
```
Comment (jcouv) [inline src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs]: I think `analyzerManager` can be declared at line 660, where it is used.

File: src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs [hunk 4]
```diff
@@ -559,304 +571,337 @@ private int RunCore(TextWriter consoleOutput, ErrorLogger errorLogger, Cancellat
...
+            NoThrowStreamDisposer sourceLinkStreamDisposerOpt = null;
+
+            try
+            {
+                // NOTE: Unlike the PDB path, the XML doc path is not embedded in the assembly, so we don't need to pass it to emit.
+                var emitOptions = Arguments.EmitOptions.
+                    WithOutputNameOverride(outputName).
+                    WithPdbFilePath(PathUtilities.NormalizePathPrefix(finalPdbFilePath, Arguments.PathMap));
+
+                // This feature flag is being maintained until our next major release to avoid unnecessary
```
Comment (jcouv) [inline src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs]: >next major releas [](start = 67, length = 17)

This is not introduced by this PR, but which release is the "next major release". Was it 15.0? Is it 16.0?
Could you file a follow-up issue?

File: src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs [hunk 5]
```diff
@@ -1018,64 +1063,65 @@ protected virtual string GetOutputFileName(Compilation compilation, Cancellation
...
 
-        protected Stream GetWin32Resources(CommandLineArguments arguments, Compilation compilation, out IEnumerable<DiagnosticInfo> errors)
+        protected Stream GetWin32Resources(CommandLineArguments arguments, Compilation compilation, DiagnosticBag diagnostics)
         {
-            return GetWin32ResourcesInternal(MessageProvider, arguments, compilation, out errors);
+            return GetWin32ResourcesInternal(MessageProvider, arguments, compilation, diagnostics);
         }
 
         // internal for testing
         internal static Stream GetWin32ResourcesInternal(CommonMessageProvider messageProvider, CommandLineArguments arguments, Compilation compilation, out IEnumerable<DiagnosticInfo> errors)
```
Comment (jcouv) [inline src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs]: Nit: I think we could change IEnumerable to ImmutableArray in `GetWin32ResourcesInternal` and `GetWin32Resources` signatures too

File: src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs (hunk 2 above)
Comment (agocke) [reply src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs]: They still use `IEnumerable` unfortunately.

File: src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs (hunk 4 above)
Comment (agocke) [reply src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs]: I believe that is this issue: https://github.com/dotnet/roslyn/issues/19592

**Pull request merged**